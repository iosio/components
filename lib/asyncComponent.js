(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', './FadeSwitch'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('./FadeSwitch'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.FadeSwitch);
        global.asyncComponent = mod.exports;
    }
})(this, function (exports, _react, _FadeSwitch) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.asyncComponent = undefined;

    var _react2 = _interopRequireDefault(_react);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    var asyncComponent = exports.asyncComponent = function asyncComponent(getComponent, Loader, duration, fadeSwitch_props, fader_props) {
        var _class, _temp;

        return _temp = _class = function (_React$Component) {
            _inherits(AsyncComponent, _React$Component);

            function AsyncComponent() {
                _classCallCheck(this, AsyncComponent);

                var _this = _possibleConstructorReturn(this, (AsyncComponent.__proto__ || Object.getPrototypeOf(AsyncComponent)).call(this));

                _this.state = { Component: AsyncComponent.Component };

                if (!_this.state.Component) {
                    getComponent().then(function (Component) {
                        Component = Component && Component.__esModule ? Component.default : Component;
                        AsyncComponent.Component = Component;
                        _this.setState({ Component: Component });
                    });
                }
                return _this;
            }

            _createClass(AsyncComponent, [{
                key: 'render',
                value: function render() {
                    var Component = this.state.Component;


                    return _react2.default.createElement(
                        _react2.default.Fragment,
                        null,
                        duration > 0 ? _react2.default.createElement(_FadeSwitch.FadeSwitch, Object.assign({}, fadeSwitch_props, {
                            fader_props: fader_props,
                            duration: duration
                        }, this.props, {
                            style: { height: '100%', width: '100%' },
                            view: !!Component ? 1 : 0,
                            View0: Loader ? _react2.default.createElement(Loader, null) : null,
                            View1: Component ? _react2.default.createElement(Component, this.props) : null
                        })) : Component ? _react2.default.createElement(Component, this.props) : null
                    );
                }
            }]);

            return AsyncComponent;
        }(_react2.default.Component), _class.Component = null, _temp;
    };
});