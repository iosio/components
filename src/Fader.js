import React from 'react';
export class Fader extends React.PureComponent {
    render() {
        const { className, style, show, children, duration, ref, ...rest } = this.props;
        const value = duration ? duration : 175;
        return (
            <div
                id="Fader"
                ref={ref}
                className={className}
                style={{
                    ...style,
                    opacity: show ? 1: 0,
                    transition: `opacity ${value}ms ease-in-out`
                }}
                {...rest}>

                {children}
            </div>
        );
    }
}









