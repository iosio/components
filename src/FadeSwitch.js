import React from 'react';

import {Fader} from "./Fader";

export class FadeSwitch extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            show: false,
            view: 0,
            initial_load: true,
        };
        this.is_mounted_ = false;
        this.timeout = false;
        this.duration = props.duration || props.duration === 0 ? props.duration : 175;
    }

    componentDidMount() {
        this.is_mounted_ = true;
        this.showView(this.props.view);
    }

    showView = (view_num) => {

        if (this.is_mounted_) {

            if (!this.state.initial_load) {

                this.setState({show: false});

                this.timeout = setTimeout(() => {
                    if (this.is_mounted_) {
                        this.setState({view: view_num, show: true,});
                    }
                }, this.duration);

            } else {
                this.setState({view: view_num, show: true, initial_load: false});
            }
        }
    };

    componentDidUpdate(prevProps) {
        if (prevProps.view !== this.props.view) {
            this.showView(this.props.view);
        }
    }

    componentWillUnmount() {
        this.is_mounted_ = false;
        clearTimeout(this.timeout);
    }

    render() {
        const {
            View0,
            View1,
            className,
            fader_props,
            style
        } = this.props;

        const {
            show,
            view,
        } = this.state;

        let fp_style = {};
        if (fader_props && fader_props.style) {
            fp_style = fader_props.style;
        }
        const combined_style = {...style, ...fp_style};

        return (
            <Fader
                duration={this.duration}
                className={className}
                style={{
                    height: '100%',
                    width: '100%',
                    position: 'relative',
                    ...combined_style,
                }}
                show={show}>

                {view === 0 ? View0 : View1}
            </Fader>

        );
    };
}
