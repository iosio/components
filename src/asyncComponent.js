import React from 'react';
import {FadeSwitch} from "./FadeSwitch";

export const asyncComponent = (getComponent, Loader, duration, fadeSwitch_props, fader_props) => {
    return class AsyncComponent extends React.Component {
        constructor() {
            super();
            if (!this.state.Component) {
                getComponent().then(Component => {
                    Component = Component && Component.__esModule ? Component.default : Component;
                    AsyncComponent.Component = Component;
                    this.setState({Component});
                });
            }
        }

        static Component = null;
        state = {Component: AsyncComponent.Component};

        render() {
            const {Component} = this.state;

            return (
                <React.Fragment>
                    {Loader && duration > 0 ? <FadeSwitch
                            {...fadeSwitch_props}
                            fader_props={fader_props}
                            duration={duration}
                            {...this.props}
                            style={{height: '100%', width: '100%'}}
                            view={!!Component ? 1 : 0}
                            View0={Loader ? <Loader/> : null}
                            View1={Component ? <Component {...this.props}/> : null}
                        />
                        :

                        (Component ? <Component {...this.props}/> : null)
                    }
                </React.Fragment>
            );
        }
    }
};


