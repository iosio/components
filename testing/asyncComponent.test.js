import React from 'react';
import renderer from 'react-test-renderer';

import {asyncComponent} from "../src/asyncComponent";


const wait = time => new Promise(resolve => setTimeout(resolve, time));

const MyComponent = (props) => <div>MyComponent props: {JSON.stringify(props)}</div>;
const MyLoadingComponent = () => <div>MyLoadingComponent</div>;


function createLoader(delay, loader, error) {
    return () => {
        return wait(delay).then(() => {
            if (loader) {
                return loader();
            } else {
                throw error;
            }
        });
    };
}


describe('asyncComponent', () => {

    it('should successfully load', async () => {

        const fade_duration = 1;

        const MyAsyncComp = asyncComponent(
            createLoader(1, () => MyComponent), //component to load
            MyLoadingComponent, //loader/loading animation
            fade_duration //
        );

        const tree = renderer.create(
            <MyAsyncComp prop={'foo'}/>
        );

        expect(tree).toMatchSnapshot(); //loading

        await wait(1);

        expect(tree).toMatchSnapshot(); //loaded


    });


});