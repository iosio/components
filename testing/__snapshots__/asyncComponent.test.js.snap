// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncComponent should successfully load 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 1ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div>
    MyLoadingComponent
  </div>
</div>
`;

exports[`asyncComponent should successfully load 2`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 1ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div>
    MyComponent props: 
    {"prop":"foo"}
  </div>
</div>
`;
